### ISSUE TITLE
Implement a Type Builder

### ISSUE THREAD

**Issue by treb93** [easyrdf/easyrdf#361]
Hello, 

As there's already a Type Mapper implemented (resources of type foaf:Person can be mapped into PHP object of class Foaf\Person), I wonder if you already figure out how to develop a Type Builder ?

The idea is to be able to generate classes and HTML forms, from any RDF schema.

I need this for a project and I could help / take in charge some devs if needed

**Comment by k00ni** [comment]
CC @njh

**Comment by k00ni** [comment]
Hello, I am not sure what the current state of the Type Builder is, but @njh can tell more about it.

**From a maintainer perspective:** We got EasyRdf back on track to be maintained again(baby steps). Although I like new things, we should keep that in mind (someone has to maintain it). Lets assume EasyRdf's Type Builder is not sufficient enough, can you tell us about the extend of the development you are planning? 

Is it possible that you provide (sufficient) test coverage and some documentation besides the code itself? That would be really great, but I am not sure if its not too much of an ask.

**Comment by treb93** [comment]
Yes.

I intend to develop an Open Source solution to generate HTML forms / Objects types from any RDF schema & save it in a triplestore. 

With Symfony it seems quite easy to generate Forms from Object classes, EasyPHP provides mapping to output rdf, so the last step to implement is to be able to create an object type from a rdf schema.
I also plan to use ARC2 for the database storage 

I can provide some doc / test coverage for the devs I take charge.

**Comment by zozlak** [comment]
Please forgive me my scepticism but I would propose to implement such a feature as a separate library.

There are many decisions to be made while implementing such a feature. Just for starters:

* What does "RDF schema" means? An OWL ontology providing class definitions or maybe something ad-hoc which is driven with data? Or maybe something hand-crafted for the purpose of this feature (more straightforward to use then OWL but providing a fixed class definition in contrary to the ad-hoc data-driven mapping)?
* How to map RDF property URIs to something object oriented programmers friendly?

Such decisions seem to me out of scope of the EasyRdf and I would definitely prefer to have a clear API separation between the EasyRdf core and such a mapper. And if we agree on such a separation, moving such a mapper to a separate library will only make it easier to develop both in parallel (especially it will assure we don't need to sync development of both).

**Comment by k00ni** [comment]
@zozlak made some good points.

Will your project rely on further (Symfony?) components? If so, that would be a problem in my opinion, because EasyRdf should not get further dependencies for (kinda) optional components. That approach is also in line with Symfony's, creating separate repositories for certain use cases.

:+1: for the project, but unfortunately :-1: on doing it in this repository.

I also suggest that you create your own repository, so you can start right away and don't have to wait on our opinions about your project. In case you wanna hear mine I would be delighted for a mention in issues or PRs. When your project reaches maturity  we can discuss about moving it to the easyrdf organization, if you still want that. For me, that only would makes sense, if you are tight to EasyRdf classes and functions though.

**Comment by treb93** [comment]
Alright, I'll open something on gitlab and you'll see later if you want to include it to the project. 

FYI, by "Rdf Schema" I mean vocabularies like Schema.org or Foaf which structure is downloadable in Rdf. (I confess I'm quite a newbie in Web semantics ;) )
I don't intend to rely on Symfony components for this component.

Be in touch !

**Comment by zozlak** [comment]
I guess it will be more difficult than you expect...

If you look at schema.org, you will realize they are using their very own schema to describe their schema :) But at least their very own schema is simple and straightforward. On the other hand their schema for describing schemas doesn't provide a way for denoting cardinality restrictions (e.g. "class A must have exactly one value of property P" or "class A may have 0 or 1 values of property P") which might or might not be an important feature for your use case.

If you take a look at FOAF, you will see they are describing it with OWL. Which on one hand is good because it's an W3C standard but on the other OWL is extremely complex and doesn't really match your use case (OWL tells what you can infer from data and not how data have to be structured to be considered valid and, surprisingly, it quite often makes a significant difference).

Property range handling will be another big issue. Let's say we want to automatically build a form for a schema in which *an object of class A must have a property P denoting an object of class B*. But what should user really provide in the form as the P property value? An URI? But can we expect (s)he know the URI and how we can check if the entity with a user-provided URI is really of class B? Or maybe the user should select something from a drop-down list? But how we can obtain a list of all objects of class B and what we should display to the user so (s)he can choose the desired one (because URIs might not be meaningful enough)?

Please don't get me wrong, I don't want to disencourage you from implementing your idea. I just wanted to stress that the semantic world is pretty unintuitive and implementing ORM concepts in it can be pretty complex.

**Comment by k00ni** [comment]
Will close this for now due to no activity. Feel free to reopen.
